### DISCUSSION TITLE
💙 F# の好きなところについて教えてください！

### DISCUSSION THREAD

**Discussion by fffsharp** [fffsharp/Forum#5] [雑談]
Fun Fan F# の資料作成のヒントにしたいので、ご協力いただけますと幸いです🙇‍♂️

**Comment by sayurin** [comment]
左結合に偏った演算子の優先順位と`|>`とにより、上から下、左から右、へと流れる記述ができるところ。更に型推論によりほとんど型を記述しなくても良い点。トカ？

ちなみに
```
let y = x
z
```
と書くと特に`x`が長いと読み辛くなるので

```
x
|> fun y -> z
```

とやったりもします。

**Comment by joemphilips** [comment]
###  1. 気軽に一時スコープを切れる

```fsharp
let x = 1
let y = 2
let z = x + y
```

と書く代わりに
```fsharp
let z =
  let x = 1
  let y = 2
  x + y
```

と書くことで `x` , `y` が一時変数であることが明らかになり、保守しやすい

### 2. ファイル間の循環参照が不可能

型/変数間の依存関係がひと目で把握できるのでやはり保守しやすい

### 3. Object 式

```fsharp
let i = { new IDisposable with member this.Dispose() = failwith "todo" }
```

のように、 特定の inteface を実装した型をその場でインスタンス化できる。
特にテストで便利。 mock 系ライブラリが必要なくなる。

**Reply by sayurin** [reply]
つい先日オブジェクト式が活躍する場面がありました。具体的には[JSONシリアライザのカスタムコンバーター](https://docs.microsoft.com/ja-jp/dotnet/standard/serialization/system-text-json-converters-how-to?pivots=dotnet-6-0)をインラインで実装しました。使用したい個所で直接定義できるのはとても良いです。

```F#
let options = JsonSerializerOptions(PropertyNamingPolicy = JsonNamingPolicy.CamelCase)
options.Converters.Add {
    new JsonConverter<Position>() with
        member _.Read(reader, typeToConvert, options) = NotImplementedException() |> raise
        member _.Write(writer, value, options) = writer.WriteStringValue $"{value}"
}
```

ちなみにオーバーライドする以上、メソッド引数は自明なため、ここも型推論されます。

ちなみにちなみに、C#同様、オブジェクト初期化子はあるので`PropertyNamingPolicy`の初期化はできますが、コレクション初期化子がないので`Converters`は明示的に`Add()`を呼ぶ必要があります。C#だと

```C#
var options = new JsonSerializerOptions {
    PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
    Converters = { ... },
}
```

と書けます。F#にもほすぃ…。

**Reply by htsign** [reply]
@sayurin 適当に作ったものですが、近いことはできるかなと思います。どうでしょう？

```fsharp
let inline withInitCollection<'m, ^a, 'b, 'c when ^a : (member Add : 'b -> 'c)> (selector : 'm -> ^a) (xs : 'b seq) (m : 'm) =
    let add x xs = (^a : (member Add : 'b -> 'c) (xs, x))
    let collection = selector m
    Seq.iter (fun x -> add x collection |> ignore) xs
    m
```

```fsharp
let options =
    JsonSerializerOptions(PropertyNamingPolicy = JsonNamingPolicy.CamelCase)
    |> withInitCollection (fun x -> x.Converters) [ ... ]
```

**Comment by tatsuya-midorikawa** [comment]
1. Unit of Measure
すべての言語に入って欲しい機能

2. (TBD)

**Comment by wraikny** [comment]
* SRTP（抽象化！）
* コンピューテーション式（書きやすい！）
* TypeProvider（お手軽！）
* .NETの資産を利用できる（嬉しい！）

**Comment by murata2makoto** [comment]
Active Pattern

XMLの構文解析結果を、分かりやすいパターンマッチで処理できます。自前のデータ構造にあらかじめ変換しておく必要がありません。

実例: パターンマッチの実例が[これ](https://github.com/CITPCSHARE/CREPDL/blob/Version2/CREPDL/StringValidation.fs)で、Active Patternの定義が[これ](https://github.com/CITPCSHARE/CREPDL/blob/Version2/CREPDL/ActivePattern.fs)ね。
